service/matching/tasklist.go

```diff
@@ -0,0 +1,150 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package matching
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/uber/cadence/common/persistence"
+)
+
+type (
+	// taskListID is the key that uniquely identifies a task list
+	taskListID struct {
+		qualifiedTaskListName
+		domainID string
+		taskType int
+	}
+	// qualifiedTaskListName refers to the fully qualified task list name
+	qualifiedTaskListName struct {
+		name      string // internal name of the tasks list
+		baseName  string // original name of the task list as specified by user
+		partition int    // partitionID of task list
+	}
+)
+
+const (
+	// taskListPartitionPrefix is the required naming prefix for any task list partition other than partition 0
+	taskListPartitionPrefix = "/__cadence_sys/"
+)
+
+// newTaskListName returns a fully qualified task list name.
+// Fully qualified names contain additional metadata about task list
+// derived from their given name. The additional metadata only makes sense
+// when a task list has more than one partition. When there is more than
+// one partition for a user specified task list, each of the
+// individual partitions have an internal name of the form
+//
+//     /__cadence_sys/[original-name]/[partitionID]
+//
+// The name of the root partition is always the same as the user specified name. Rest of
+// the partitions follow the naming convention above. In addition, the task lists partitions
+// logically form a N-ary tree where N is configurable dynamically. The tree formation is an
+// optimization to allow for partitioned task lists to dispatch tasks with low latency when
+// throughput is low - See https://github.com/uber/cadence/issues/2098
+//
+// Returns error if the given name is non-compliant with the required format
+// for task list names
+func newTaskListName(name string) (qualifiedTaskListName, error) {
+	tn := qualifiedTaskListName{
+		name:     name,
+		baseName: name,
+	}
+	if err := tn.init(); err != nil {
+		return qualifiedTaskListName{}, err
+	}
+	return tn, nil
+}
+
+// IsRoot returns true if this task list is a root partition
+func (tn *qualifiedTaskListName) IsRoot() bool {
+	return tn.partition == 0
+}
+
+// Parent returns the name of the parent task list
+// input:
+//   degree: Number of children at each level of the tree
+// Returns empty string if this task list is the root
+func (tn *qualifiedTaskListName) Parent(degree int) string {
```

Comment:
`degree` must be > 1

Reply:
Also, shouldn't degree be configured on the task list.. Why is this an input to the method?

Reply:
if its not an input, then this type needs to take in the dynamic config func as input in the constructor. the latter option results in passing dynamic config everywhere in the unit test. But the actual Parent() method is only going to be called in one or two places. So, I  decided to pass this a param to this function.